### ISSUE TITLE
Moving Columns with DEFAULT NULL doesn't work on MariaDB 10.2+

### ISSUE THREAD

**Issue by adwiv** [phpmyadmin/phpmyadmin#14951]
### Describe the bug
When a column is nullable with default value NULL, trying to move it using the "Move Columns" popup either does not work or works incorrectly with disastrous side effects. 

It appears that the SQL generated for **moving the column converts the NULL keyword value into 'NULL' string value**. If the column is a text column, the default value is changed to 'NULL' String instead of the NULL VALUE. If the column is some other type (eg int), the column moving fails with an error since the string 'NULL' is not a valid integer value.

### To Reproduce
Create a simple table
```sql
CREATE TABLE test (
 col1 int(11) DEFAULT NULL,
 col2 varchar(20) DEFAULT NULL
)
```
Click on the `Structure` tab and select "Move Columns". Try to move col2 to top of table. The SQL generated is:

```sql
ALTER TABLE `test` CHANGE `col2` `col2` VARCHAR(20) NULL DEFAULT 'NULL'  FIRST
```
Notice the 'NULL' in quotes which changes the default value of the column.

Now try moving column col1 to top of table again. This generates the following SQL
```sql
ALTER TABLE `test` CHANGE `col1` `col1` INT(11) NULL DEFAULT 'NULL' FIRST
```
This time running it fails with error ` #1366 - Incorrect integer value: 'NULL' for column 'col1' at row 1`

### Expected behavior

Columns should have moved without any side effects. The default value should have remained NULL, and not converted to string 'NULL'

### Server configuration

- MAC OS X 10.14.2
- Web server: Apache/2.4.35 (Unix)
- Database version: mysqlnd 5.0.12
- PHP version: 7.3.1
- phpMyAdmin version: 4.8.5 and 5.0+snapshot

**Comment by williamdes** [comment]
Tested on demo servers (root, no password)
- [`5.0.0-dev`](https://demo.phpmyadmin.net/master-config/) (not affected)
- [`4.8.x`](https://demo.phpmyadmin.net/QA_4_8/) (affected)

![QA_4_8](https://user-images.githubusercontent.com/7784660/52663936-1af4b780-2f08-11e9-9051-29c35811e9f5.png)


On `QA_4_8` there is a modal that is broken.
Can not test more.

**Comment by adwiv** [comment]
On further testing I can confirm that the bug is somehow dependent on Server version.

The latest snapshot works fine with MariaDB 10.1.38, but fails with newer MariaDB 10.2.22 and 10.3.12, whereas MySQL 5.7/8.0 do not seem to be affected. 

Tested using docker images on Ubuntu 18.04 with apache and phpMyAdmin-5.0+snapshot-all-languages.tar.gz

**Comment by kartik1000** [comment]
@williamdes, Hey it's working fine for me both on mariadb and mySql Server on demo server. I'll share the screenshot with you and in preview Sql mode also its working fine for me 
![Screenshot 2019-03-10 at 1 09 50 AM](https://user-images.githubusercontent.com/43112279/54076441-ab3cd700-42d1-11e9-9e5a-2f0bec779209.png)

**Comment by williamdes** [comment]
@kartik1000 Can you please try on `wdes`'s demo servers ? (we have enough mariadb and mysql servers)

https://phpmyadmin-pr-6.wdes.eu.org and https://phpmyadmin-pr-7.wdes.eu.org

public/public

**Comment by kartik1000** [comment]
@williamdes, Hey Yes You are right It ain't working on these servers switching col 1 to top does give back an error and the error is same as stated by @adwiv , and moreover  https://phpmyadmin-pr-7.wdes.eu.org does not show sql preview option whereas the other one does.

**Comment by williamdes** [comment]
Thank you @kartik1000 :)

**Comment by williamdes** [comment]
I could not reproduce this issue on demo servers (`root`, no password)
- [`5.0.0-dev`](https://demo.phpmyadmin.net/master/)
- [`4.9.x`](https://demo.phpmyadmin.net/QA_4_9/)

**Comment by adwiv** [comment]
@williamdes The demo servers referred by you use MariaDB 10.1.26 whereas the errors happen with newer 10.2 or 10.3.

**Comment by williamdes** [comment]
@adwiv please try on my demo server https://phpmyadmin-pr-7.wdes.eu.org/ (QA_4_9)

The servers are `MariaDB 1 > 10.3` `MariaDB 2 > 10.2` `MariaDB 3 > 10.3`

User `root` or `public` and password `public`

I still can not reproduce the issue

**Comment by williamdes** [comment]
![image](https://user-images.githubusercontent.com/7784660/62836205-13b52900-bc61-11e9-9724-abbfae4832e5.png)
```
#1366 - Incorrect integer value: 'NULL' for column ``.``.`col1` at row 1
```

**Comment by adwiv** [comment]
> Reproduced: you need to move col2 to first position and then try to move it back !

Yes, if the column type is VARCHAR, the move changes NULL default value to string "NULL". 
Second time, the error is shown because INT column is moved as string "NULL" cant be converted to int.

If you use the following schema, it will show the error everytime.

```sql
CREATE TABLE test (
 col1 int(11) DEFAULT NULL,
 col2 varchar(20) DEFAULT NULL
)
```
In short, the code is somehow interpreting NULL default value as string "NULL". In case of any non-text column, it will show error. And in case of text column, it won't show error but will corrupt the schema.

**Comment by adwiv** [comment]
Okay. I figured out the cause of issue. 

Starting with MariaDB 10.2, they have made some changes to the Information Scheme Table which deviates from MySQL. [ChangeLog Here](https://mariadb.com/kb/en/library/changes-improvements-in-mariadb-102/#information-schema)

> Changes to the Information Schema `COLUMNS` table. **Literals are now quoted** in the `COLUMN_DEFAULT` column to distinguish them from expressions ([MDEV-13132](https://jira.mariadb.org/browse/MDEV-13132)), and two new columns added providing information about generated (virtual, or computed) columns ([MDEV-9255](https://jira.mariadb.org/browse/MDEV-9255)).

So, when we have a nullable column with NULL default in MariaDb, instead of getting a null value from the information_schema table for COLUMN_DEFAULT, we get a literal string with value NULL.

**Comment by kartik1000** [comment]
@adwiv, you must be right because I checked the changes made in the codebase when moving columns, and checked if what change possibly triggered this issue as it was working fine earlier but couldn't find any. @williamdes what can we now do possibly to fix the issue?

**Comment by williamdes** [comment]
@kartik1000 I would add an override to the existing code

**Comment by kartik1000** [comment]
@kartik1000 It will implement the desired code after checking the MariaDB version right?

**Comment by williamdes** [comment]
@kartik1000 Yes, perfect :)

**Comment by adwiv** [comment]
@kartik1000 @williamdes There is another problem introduced due to a recent commit https://github.com/phpmyadmin/phpmyadmin/commit/98a7479174a081ac60419034151fd89ec952265f which results in column always becoming NOT NULL when moved. This commit changed the expected value of `$null` parameter in function `Table::generateFieldSpec` from NULL/NOT NULL to YES/NO. 

https://github.com/phpmyadmin/phpmyadmin/blob/d3ef2ce41dba393c0d7d8fdc9cd89a38d5420908/libraries/classes/Table.php#L559-L563

However, the expected value was not not updated in `TableStructure::moveColumns` function which is still passing NULL/NOT NULL as parameters.
https://github.com/phpmyadmin/phpmyadmin/blob/d3ef2ce41dba393c0d7d8fdc9cd89a38d5420908/libraries/classes/Controllers/Table/StructureController.php#L464

So, in case of moving a nullable field, `Table::generateFieldSpec` receives parameter $null as 'NULL' but is expecting 'YES', so it outputs 'NOT NULL' as field definition. The resolution is to simply change the above line to send parameters YES/NO instead of NULL/NOT NULL.

Thanks.

**Comment by williamdes** [comment]
@adwiv Thank you so much for your comment!
Making the fix soon

**Comment by williamdes** [comment]
Hi @adwiv thanks for your patience !
can you try to apply 8aad190ae035ec36db9376375473c709c6341fb0 ?

You can also add e9ad2bc153c8e7c6a079aa6f12ddca590451d40c

**Comment by Kaulkwappe** [comment]
Good evening! Was this problem fixed? Because it still concerns me: 

> #1366 - Incorrect integer value: 'NULL' for column ``.``.`last_login_time` at row 1

```
Apache/2.4.38
PHP 7.3.11-1~deb10u1
10.3.18-MariaDB-0+deb10u1
phpMyAdmin 4.6.6deb4
```

**Comment by williamdes** [comment]
@Kaulkwappe your phpmyadmin version is very out to date
As you can see in #15236 phpMyAdmin 4.9 will be shipped in the next debian version and for now you can use our Ubuntu ppa #15515 

The issue seems is fixed in latest versions

**Comment by Kaulkwappe** [comment]
I see, thanks for the fast response. Seems I have to wait since 4.6.6deb4 is the most recent available version in Debian Buster 10.2.
